File: Merge_Sort.java
import java.util.*;

public class MergeSort {
  public static void main(String[] args) {
    Scanner sc = new Scanner(System.in);
    System.out.println("Enter the size of the Array: ");
    int size_of_arry = sc.nextInt(); // size of array, to b entered by the user
    int arr[] = new int[size_of_arry];
    System.out.println("Enter Array Elements ");
    for (int i = 0; i < size_of_arry; i++)
      arr[i] = sc.nextInt(); // element of array, to be entered by the user
    arr = Merge_sort(arr, size_of_arry);
    System.out.println("Array after Merge Sort is: ");
    for (int i = 0; i < size_of_arry; i++)
      System.out.print(arr[i] + " ");
    System.out.println("\n");
  }

  static int[] Merge_sort(int arr[], int size) {
    if (size > 1) {
      int mid = size / 2;
      int[] first = Arrays.copyOfRange(arr, 0, mid);
      first = Merge_sort(first, mid); // recursive call for first half array
      int[] second = Arrays.copyOfRange(arr, mid, size);
      second = Merge_sort(second, size - mid); // recursive call for second half array
      arr = Merge_arrays(first, second, mid, size - mid);
    }
    return arr;
  }

  static int[] Merge_arrays(int first[], int second[], int n, int m) // respectively
  {
    int arr[] = new int[n + m];
    int i = 0, f = 0, s = 0;
    while (f < n && s < m) {
      if (first[f] < second[s])
        arr[i++] = first[f++];
      else
        arr[i++] = second[s++];
    }
    while (f < n)
      arr[i++] = first[f++];
    while (s < m)
      arr[i++] = second[s++];
    return arr;
  }
}

File: mergesort.cpp
#include <bits/stdc++.h>
using namespace std;

void merge(int arr[], int l,int mid, int h)
{
    int n_l = mid-l+1, n_r = h-mid;
    int arr_l[n_l], arr_r[n_r];
    
    for(int i=0; i<n_l; i++)
    {
        arr_l[i] = arr[l+i]; 
    }
    
    for(int j=0; j<n_r; j++)
    {
        arr_r[j] = arr[mid+j+1];
    }
    
    int i=0,j=0,k=l;
    
    while(i<n_l && j<n_r)
    {
        if(arr_l[i]<=arr_r[j])
        {
            arr[k]=arr_l[i];
            i++;
        }
        else
        {
            arr[k]=arr_r[j];
            j++;
        }
        k++;
    }
    
    while(i<n_l)
    {
        arr[k]=arr_l[i];
        i++;
        k++;
    }
    while(j<n_r)
    {
        arr[k]=arr_r[j];
        j++;
        k++;
    }
}

void mergesort(int arr[], int l, int h)
{
    if(l<h)
    {
        int mid = (l+h)/2;
        mergesort(arr,l,mid);
        mergesort(arr,mid+1,h);
        merge(arr,l,mid,h);
    }
}

int main()
{
    int arr[10] = {34,12,45,2,89,67,64,589,21,54};
    mergesort(arr,0,9);
    for(int i=0; i<10; i++)
    {
        cout<<arr[i]<<" ";
    }
}